src/dbnode/storage/index/mutable_segments.go

```diff
@@ -0,0 +1,795 @@
+		segs = append(segs, compaction.Segment{
+			Age:     seg.Age(),
+			Size:    seg.Segment().Size(),
+			Type:    segments.FSTType,
+			Segment: seg.Segment(),
+		})
+	}
+
+	plan, err := compaction.NewPlan(segs, m.opts.BackgroundCompactionPlannerOptions())
+	if err != nil {
+		instrument.EmitAndLogInvariantViolation(m.iopts, func(l *zap.Logger) {
+			l.Error("index background compaction plan error", zap.Error(err))
+		})
+		return
+	}
+
+	if len(plan.Tasks) == 0 {
+		return
+	}
+
+	// Kick off compaction.
+	m.compact.compactingBackground = true
+	go func() {
+		m.backgroundCompactWithPlan(plan)
+
+		m.Lock()
+		m.compact.compactingBackground = false
+		m.cleanupBackgroundCompactWithLock()
+		m.Unlock()
+	}()
+}
+
+func (m *mutableSegments) shouldEvictCompactedSegmentsWithLock() bool {
+	// NB(r): The frozen/compacted segments are derived segments of the
+	// active mutable segment, if we ever evict that segment then
+	// we don't need the frozen/compacted segments either and should
+	// shed them from memory.
+	return m.state == mutableSegmentsStateClosed ||
+		m.hasEvictedAnyTimes
+}
+
+func (m *mutableSegments) cleanupBackgroundCompactWithLock() {
+	if m.state == mutableSegmentsStateOpen {
+		// See if we need to trigger another compaction.
+		m.maybeBackgroundCompactWithLock()
+		return
+	}
+
+	// Check if need to close all the compacted segments due to
+	// having evicted mutable segments or the block being closed.
+	if !m.shouldEvictCompactedSegmentsWithLock() {
+		return
+	}
+
+	// Evict compacted segments.
+	m.closeCompactedSegments(m.backgroundSegments)
+	m.backgroundSegments = nil
+
+	// Free compactor resources.
+	if m.compact.backgroundCompactor == nil {
+		return
+	}
+
+	if err := m.compact.backgroundCompactor.Close(); err != nil {
+		instrument.EmitAndLogInvariantViolation(m.iopts, func(l *zap.Logger) {
+			l.Error("error closing index block background compactor", zap.Error(err))
+		})
+	}
+	m.compact.backgroundCompactor = nil
+}
+
+func (m *mutableSegments) closeCompactedSegments(segments []*readableSeg) {
+	for _, seg := range segments {
+		err := seg.Segment().Close()
+		if err != nil {
+			instrument.EmitAndLogInvariantViolation(m.iopts, func(l *zap.Logger) {
+				l.Error("could not close compacted segment", zap.Error(err))
+			})
+		}
+	}
+}
+
+func (m *mutableSegments) backgroundCompactWithPlan(plan *compaction.Plan) {
+	sw := m.metrics.backgroundCompactionPlanRunLatency.Start()
+	defer sw.Stop()
+
+	n := m.compact.numBackground
+	m.compact.numBackground++
+
+	logger := m.logger.With(
+		zap.Time("blockStart", m.blockStart),
+		zap.Int("numBackgroundCompaction", n),
+	)
+	log := n%compactDebugLogEvery == 0
+	if log {
+		for i, task := range plan.Tasks {
+			summary := task.Summary()
+			logger.Debug("planned background compaction task",
+				zap.Int("task", i),
+				zap.Int("numMutable", summary.NumMutable),
+				zap.Int("numFST", summary.NumFST),
+				zap.String("cumulativeMutableAge", summary.CumulativeMutableAge.String()),
```

Comment:
Know that this is existing code, but quickly - can use `zap.Stringer("foo", bar)` so these debug logs do not need to call ".String()" each time (unless debug logs actually turned on).